### PR TITLE
Final- Rotem

### PR DESCRIPTION
**Pull request by rotemorad**

### CODE REVIEW

File: hw5.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    """
+    Reads and analyzes data generated by the questionnaire experiment.
+    Should be able to accept strings and pathlib.Path objects.
+    """
+
+    def __init__(self, data_fname: Union[pathlib.Path, str]):
+        try:
+            self.data_fname = pathlib.Path(data_fname).resolve()
+        except TypeError("Please provide a path"):
+            raise
```
Comment (ZviBaratz) [inline hw5.py:18]: Should be:
```python
    except TypeError:
        raise TypeError("Please provide a path.")
```

File: hw5.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        hist, bins = np.histogram(self.data["age"].dropna(), bins=np.linspace(0, 100, 11))
+        self.data.hist(column="age", bins=bins)
+        return hist, bins
+
+    def remove_rows_without_mail(self) -> pd.DataFrame:
+        """Checks self.data for rows with invalid emails, and removes them.
+
+    Returns
+    -------
+    df : pd.DataFrame
```
Comment (ZviBaratz) [inline hw5.py:50]: Mind your indentation!

File: hw5.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+      A corrected DataFrame, i.e. the same table but with the erroneous rows removed and
+      the (ordinal) index after a reset.
+        """
+        email_pattern = self.data[
+            self.data["email"].str.match(r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)")]  # Using RegEx
```
Comment (ZviBaratz) [inline hw5.py:55]: Please consider:
```python
pattern = r"(^[a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+$)"
mask = self.data["email"].str.match(pattern)
return self.data[mask].reset_index(drop=True)
```

File: hw5.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+
+    Returns
+    -------
+    df : pd.DataFrame
+      The corrected DataFrame after insertion of the mean grade
+    arr : np.ndarray
+          Row indices of the students that their new grades were generated
+        """
+        q_columns = self.data.filter(items=['q1', 'q2', 'q3', 'q4', 'q5'])
+        return self.data.fillna(self.data.mean()['q1':'q5']), self.data[q_columns.isnull().any(axis=1)].index.to_numpy()
```
Comment (ZviBaratz) [inline hw5.py:71]: Avoid long, illegible statements. These will be impossible to understand if you even need to debug or change this function. Follow the example above with the regex pattern to give meaningful variable names and create easily comprehensible statements.

Comment (ZviBaratz) [reply hw5.py:71]: Note: the test for this function failed.

**Comment by rotemorad** [comment]
Hi, I just ran the code again using the updated tests file and it passed
successfully. Do you mind checking again please?
Thank you,
Rotem Orad


‫בתאריך יום א׳, 24 במאי 2020 ב-17:49 מאת ‪Zvi Baratz‬‏ <‪
notifications@github.com‬‏>:‬

File: hw5.py (hunk 4 above)
Comment (rotemorad) [reply hw5.py:71]: I just ran the code again using the updated tests file and it passed successfully. Do you mind checking again please? 
image

Comment (HagaiHargil) [reply hw5.py:71]: Hi Rotem,

On my machine, in the `q1` column, second row, your code returns a value of 5.183 while the real value is 5.275. I believe there are a few other mismatches in that table, but that's the first one that's visible.

File: test_qs.py [hunk 5]
```diff
@@ -73,7 +72,7 @@ def test_fillna_rows():
     q = QuestionnaireAnalysis(fname)
     q.read_data()
     _, rows = q.fill_na_with_mean()
-    assert np.array_equal(truth, rows)
+    df.equals(truth)
```
Comment (HagaiHargil) [inline test_qs.py:75]: @rotemorad This is why your tests pass - you changed the test.

**Comment by rotemorad** [comment]
This is the test Hagi updated test Hagi uploaded, so if it doesn't work- I
had no way of knowing since it passed the test on my computer

**Comment by HagaiHargil** [comment]
It's Hagai :)

Luckily we're inside a git repository, so everything has a history :)

There were about 12 hours where the test code was the way it's written above (i.e. incorrect), but I asked to pull the repo after it was fixed. Indeed there's a small chance that you pulled the first time, but never pulled after the correction, and so you remained with the erroneous test. I'll assume that this is what happened and I'll fix your grade.

**Comment by rotemorad** [comment]
Great, thank you so much!
I just want to make sure my grades are:
HW1- 92
HW2-100
HW3-100
HW4-100 (originally I got 90 but the csv file wasn't normalized properly so
you said you would update my grade to 100)
HW5- 115
Do you calculate the mean grade, including the bonus of HW5? Does it mean
my mean grade is supposed to be 100?
Thank you, and sorry for the many questions I sent during the semester. I
had a lot of fun learning python and it will be put to use in my lab in the
analysis of MRI images

Rotem Orad


‫בתאריך יום ב׳, 25 במאי 2020 ב-12:29 מאת ‪Hagai Har-Gil‬‏ <‪
notifications@github.com‬‏>:‬

**Pull request closed without merging**